Client/Frontend/Browser/MailProviders.swift

```diff
@@ -0,0 +1,192 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+
+protocol MailProvider {
+    var beginningScheme: String {get set}
+    var supportedHeaders: [String] {get set}
+    func newEmailURLFromMetadata(metadata: MailToMetadata) -> NSURL?
+}
+
+// mailto headers: subject, body, cc, bcc
+
+class ReaddleSparkIntegration : MailProvider {
+    var beginningScheme = "readdle-spark://compose?"
+    var supportedHeaders = [
+        "subject",
+        "recipient",
+        "textbody",
+        "html",
+        "cc",
+        "bcc"
+    ]
+
+    func newEmailURLFromMetadata(metadata: MailToMetadata) -> NSURL? {
+        var readdleSparkMailURL = beginningScheme
+
+        var lowercasedHeaders = [String: String]()
+        metadata.headers.forEach { (hname, hvalue) in
+            lowercasedHeaders[hname.lowercaseString] = hvalue
+        }
+
+        var toParam: String
+        if let toHValue = lowercasedHeaders["to"] {
+            let value = metadata.to.isEmpty ? toHValue : [metadata.to, toHValue].joinWithSeparator("%2C%20")
+            lowercasedHeaders.removeValueForKey("to")
+            toParam = "recipient=\(value)"
+        } else {
+            toParam = "recipient=\(metadata.to)"
+        }
+
+        var queryParams: [String] = []
+        lowercasedHeaders.forEach({ (hname, hvalue) in
+            if supportedHeaders.contains(hname) {
+                queryParams.append("\(hname)=\(hvalue)")
+            }
+
+            if hname == "body" {
+                queryParams.append("textbody=\(hvalue)")
+            }
+        })
```

Comment:
Looks like the above lines until L30 are repeated throughout each provider. Might be worth pulling this out into a private scoped function at the top of this file that returns a tuple (toParam, queryParams) to keep the code DRY.